### ISSUE TITLE
Type checking error in EQModel: heatEInWtrIM

### ISSUE THREAD

**Issue by balacij** [JacquesCarette/Drasil#3103]
Related to #3093 

This affects both 'nopcm' and 'swhs.'

```
`heatEInWtrIM` exposes ill-typed expressions!
  - Associative arithmetic operation expects all operands to be of the same type.
    Received:
      - Rational
      - Real
      - ERROR: Function application on non-function: tempW
```

There are at least 2 issues here: all 3 here in the multiplication need be of the same type, but the first 2 operands aren't the same, and the 3rd actually has an error: a function application on a non-function symbol.

https://github.com/JacquesCarette/Drasil/blob/0a240c0231b6f737011bc090962ae1925fbc83e2/code/drasil-example/swhs/lib/Drasil/SWHS/IMods.hs#L268-L283

**Comment by JacquesCarette** [comment]
So the second error is because `tempW` is typed as a `Vect Real`, but is then applied to a time value. Don't know why the temperature should be a vector. Will have to investigate that.

**Comment by JacquesCarette** [comment]
So the problem seems to be that the differential equation handling is doing some weird stuff. I think it wants a fresh variable for the system of equations, but re-using the old one instead.

**Comment by balacij** [comment]
I came across this again while trying to build out the "Beam Bending" project for @smiths' CAS 741. The design of the solution affects the underlying 'knowledge' to get code generation working. Having variables cloned and altered (as appropriate -- and perhaps only needed for this situation) seems like a good path forward.

**Comment by JacquesCarette** [comment]
There are at least three issues here:
1. adequate modelling,
2. convenient (for humans) notation,
3. context-sensitive dependence on 'stuff'.

Notation frequently elides things that can indeed safely be kept implicit. And it is context sensitive - where that context is indeed the background theory we're working in. Adequate modeling, i.e. at least well-typed and with proper units, can't do that.

So what we need (in the same order as above):
1. general models that account for reasonable things to depend on,
2. means to specify "local" notation, including what to elide,
3. ways to specialize models, for when you know you don't depend on certain things.

**Comment by smiths** [comment]
If I understand correctly @JacquesCarette you envision a general model that we specialize.  The general model accounts for things that we can depend on and when they don't matter we specialize them away.  I've always thought that way too, but I'd like to explore that thought a little more.

Let's talk about something simple like density.  To make things general, we'll talk about local density; that is, we'll focus on the density at a point inside a body:

$$\rho = \lim_{V \rightarrow 0} \frac{m}{V}$$

This is the way every physics textbook is going to write this, even though density can depend on location, temperature and pressure.  The density could also depend on time if the material is changing, maybe due to a chemical reaction.  Do we want our general model to be:

$$\rho(x, y, z, t, T, p) = \lim_{V \rightarrow 0} \frac{m(x, y, z, t, T, p}{V}$$?

For gases, we also have the function:

$$\rho = \frac{PM}{RT}$$

We have two different equations for $\rho$, but the first one, $\rho = m/V$ is the defining relation.  It is the definition for density.  It doesn't really matter where $m$ or $V$ come from, or what they depend on, this definition of density is always true.  (Similar to the definition relating energy and temperature that we discussed above.)

In cases where the defining relation is always true.  It is the definition of what that concept is, do we try to capture a general model of all the things it can depend on, or do we just capture the defining relation?

Things get more complicated when we need to move things around.  In the above, density is the dependent variable that is a function of several independent variables.  However, there will be problems where density is the independent variable.  I don't know if we can capture a functional definition of density that is always the right model for every situation.  

I think this will be easier to think about once we have a framework for our theoretical models.

**Comment by JacquesCarette** [comment]
Indeed, the 'best' general model would account for things we can depend on. However, listing all of the things it could depend on is quite unwieldy. We could handle that in two ways:
1. have some general space of parameters P that is unknown (and would have a special status as such) that we would use to mark dependence; we might need various such, since the parameter spaces don't need to all be the same in a given equation.
2. mark certain quantities as 'dependent' without saying specifically what they depend on.

An analogy would be *abstract classes* or *interfaces* in Java: they exist, we can refine them, but cannot instantiate them. One would need to always be more specific to go from such a theory to one that can be used for concrete modelling.

My guess is that the first method might formally be simpler. We can recover the second method from it by testing all quantities that depend on an abstract space of parameters as being dependent, and omit the dependency in the notation. But the modeller would be forced to pick a P when specializing, thus making their assumptions clear.
